# Predição de Severidade de Ocorrências Aeronáuticas por Tipo de Operação


Nome: Mariane Fernandes de Oliveira

Matrícula: 4052026000250

Dataset: ANAC, CENIPA

Data: Junho 2026

---

## Seção 1 — Definição do Problema

### 1.1 Contexto

O espaço aéreo brasileiro é heterogêneo: aeronaves comerciais, agrícolas e helicópteros de missões especializadas compartilham o mesmo sistema de controle de tráfego, sob regulamentações e perfis de risco fundamentalmente distintos.

O **CENIPA**, órgão do Comando da Aeronáutica, registra todas as ocorrências aeronáuticas civis desde os anos 2000, em conformidade com os normativos da ICAO [1]. O acervo — mais de treze mil ocorrências — permanece subexplorado do ponto de vista da aprendizagem de máquina. Características estruturais da operação apresentam associação documentada com a severidade dos danos [2], e este projeto parte dessa evidência para construir um modelo preditivo baseado em dados disponíveis antes da ocorrência.

### 1.2 Objetivo e Tipo de Problema

Este é um problema de **classificação multiclasse supervisionada**. A variável a ser prevista é `aeronave_nivel_dano`, com quatro categorias (NENHUM, LEVE, SUBSTANCIAL, DESTRUÍDA). O objetivo do modelo é, a partir de características da aeronave e do tipo de operação disponíveis antes da ocorrência, prever o nível de dano esperado.

**Pergunta principal:** é possível prever `aeronave_nivel_dano` com desempenho superior a um classificador de referência (baseline), a partir de características da aeronave e do tipo de operação?

**Pergunta secundária:** quais tipos de operação apresentam maior proporção de danos substanciais ou destruição total?

### 1.3 Premissas e Restrições

- Apenas variáveis disponíveis **antes** da ocorrência são usadas como feature — informações produzidas pela investigação posterior (ex.: `fator_contribuinte`) configurariam *data leakage* e são descartadas (Seção 4.1).
- `aeronave_tipo_operacao` entra no modelo como **feature**, não como target: descreve o contexto operacional em que a ocorrência se deu, e é hipótese deste projeto que ela carrega sinal relevante sobre a severidade esperada — testada diretamente na comparação entre Opção A e Opção B (Seção 5).
- O recorte temporal (ocorrências a partir de 2000, com foco analítico em 2010–2025) assume que mudanças regulatórias recentes tornam períodos muito antigos pouco representativos do risco contemporâneo (Seção 3.6).
- O desbalanceamento de classes — DESTRUÍDA representa ~4% dos registros — é tratado como característica estrutural do domínio, não como defeito a ser eliminado a qualquer custo (Seção 4.4 e Seção 6).


---
## Seção 2 — Descrição dos Dados

### 2.1 Fonte e Cobertura

Dados públicos do CENIPA/FAB disponibilizados no Portal de Dados Abertos do Governo Federal [3], cobrindo ocorrências a partir do ano 2000. A estrutura é relacional: arquivos CSV distintos cobrem aspectos diferentes de cada evento, organizados pelo identificador `codigo_ocorrencia`.

### 2.2 Tabelas Utilizadas

| Arquivo | Conteúdo | Registros (~) |
|---|---|---|
| `ocorrencia.csv` | Data, localização, classificação do evento | 13.185 |
| `aeronave.csv` | Tipo de operação, fase do voo, motor, nível de dano | 13.301 |
| `ocorrencia_tipo.csv` | Categoria taxonômica ICAO do evento | 13.900 |
| `fator_contribuinte.csv` | Fatores identificados em investigação formal | 8.613 |

`fator_contribuinte.csv` cobre apenas ~16% das ocorrências, pois fatores são documentados exclusivamente após investigação formal — reservada primariamente a acidentes graves. As implicações para a modelagem são discutidas na Seção 4.

### 2.3 Variáveis de Interesse

A seleção foi orientada por dois critérios: **cobertura** (proporção de registros não nulos) e **precedência temporal** — a variável deve existir antes da ocorrência, evitando *data leakage* [4].

| Variável | Origem | Papel |
|---|---|---|
| `aeronave_nivel_dano` | `aeronave` | **Target** |
| `aeronave_tipo_operacao` | `aeronave` | Feature |
| `ocorrencia_classificacao` | `ocorrencia` | Feature |
| `aeronave_fase_operacao` | `aeronave` | Feature |
| `aeronave_motor_tipo` | `aeronave` | Feature |
| `aeronave_motor_quantidade` | `aeronave` | Feature |
| `ocorrencia_tipo_categoria` | `ocorrencia_tipo` | Feature |

### 2.4 Variável-Alvo: `aeronave_nivel_dano`

Segue a taxonomia CENIPA/ICAO [1]: **NENHUM** (sem dano físico), **LEVE** (reparável), **SUBSTANCIAL** (dano estrutural extenso) e **DESTRUÍDA** (perda total). A distribuição empiricamente desbalanceada — DESTRUÍDA representa ~4% dos registros — constitui o principal desafio de modelagem, tratado nas Seções 4 e 6.


### 2.1 — Setup: Bibliotecas e Configurações

Antes de qualquer coisa, importamos as bibliotecas que vão sustentar todo o projeto. A escolha aqui não é arbitrária: `pandas` e `numpy` para manipulação de dados, `scikit-learn` para o pipeline de machine learning, `XGBoost` para o modelo principal, e `Plotly Express` para visualizações interativas. Matplotlib e seaborn ficam de fora — visualizações estáticas não agregam muito quando o Plotly permite inspecionar cada ponto com hover.

In [1]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    classification_report, confusion_matrix,
    accuracy_score, f1_score
)
from xgboost import XGBClassifier

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

print('Bibliotecas carregadas com sucesso.')
print(f'NumPy: {np.__version__} | Pandas: {pd.__version__}')


Bibliotecas carregadas com sucesso.
NumPy: 2.4.6 | Pandas: 3.0.3


### 2.2 — Carga dos Dados

O dataset do CENIPA chega em arquivos separados, cada um cobrindo um aspecto diferente de cada ocorrência. O trabalho aqui é juntar as peças relevantes numa estrutura única que o modelo possa consumir.

A junção principal é entre `ocorrencia.csv` (o evento) e `aeronave.csv` (a aeronave envolvida). Uma ocorrência pode envolver mais de uma aeronave — o que cria linhas duplicadas por evento. Isso é esperado e tratado durante a preparação.

In [2]:
BASE_URL = 'https://raw.githubusercontent.com/fmariane/puc-sprint2-ml/main/data/raw/CENIPA_FAB/'

ocorrencia = pd.read_csv(
    BASE_URL + 'ocorrencia.csv',
    sep=';',
    encoding='latin-1',
    low_memory=False
)

aeronave = pd.read_csv(
    BASE_URL + 'aeronave.csv',
    sep=';',
    encoding='latin-1',
    low_memory=False
)

ocorrencia_tipo = pd.read_csv(
    BASE_URL + 'ocorrencia_tipo.csv',
    sep=';',
    encoding='latin-1',
    low_memory=False
)

fator = pd.read_csv(
    BASE_URL + 'fator_contribuinte.csv',
    sep=';',
    encoding='latin-1',
    low_memory=False
)

print(f'ocorrencia:      {ocorrencia.shape}')
print(f'aeronave:        {aeronave.shape}')
print(f'ocorrencia_tipo: {ocorrencia_tipo.shape}')
print(f'fator:           {fator.shape}')


ocorrencia:      (13185, 22)
aeronave:        (13301, 23)
ocorrencia_tipo: (13900, 4)
fator:           (8613, 5)


In [3]:
# Junção ocorrencia + aeronave
df = ocorrencia.merge(
    aeronave,
    left_on='codigo_ocorrencia',
    right_on='codigo_ocorrencia2',
    how='inner'
)

# Junção com ocorrencia_tipo (pode haver múltiplos tipos por ocorrência — pegamos o primeiro)
tipo_dedup = ocorrencia_tipo.drop_duplicates(subset='codigo_ocorrencia1', keep='first')
df = df.merge(
    tipo_dedup[['codigo_ocorrencia1', 'ocorrencia_tipo_categoria', 'taxonomia_tipo_icao']],
    left_on='codigo_ocorrencia',
    right_on='codigo_ocorrencia1',
    how='left'
)

print(f'Dataset final: {df.shape[0]} linhas, {df.shape[1]} colunas')
df.head(3)
# df['aeronave_tipo_operacao'].value_counts(dropna=False)
sorted(df['aeronave_nivel_dano'].dropna().unique())


Dataset final: 13301 linhas, 48 colunas


['***', 'DESTRUÍDA', 'LEVE', 'NENHUM', 'SUBSTANCIAL']

---

## Seção 3 — Análise Exploratória dos Dados (EDA)

A EDA precede qualquer decisão de modelagem, identificando padrões estruturais que condicionam as escolhas metodológicas subsequentes [5].

### 3.1 — Visão Geral do Dataset

A inspeção inicial do dataset integrado — junção de `ocorrencia.csv` e `aeronave.csv` — revela inconsistências silenciosas: tipos incorretos inferidos pelo CSV, duplicatas por ocorrências com múltiplas aeronaves, e a convenção `***` do CENIPA para dados não informados, que impede a inferência automática de nulos pelo `pandas`.

In [4]:
print('=== Tipos de dados ===')
print(df.dtypes.value_counts())
print()
print('=== Primeiras linhas ===')
df.head(3)

=== Tipos de dados ===
str        34
int64       9
float64     5
Name: count, dtype: int64

=== Primeiras linhas ===


,codigo_ocorrencia,codigo_ocorrencia1_x,codigo_ocorrencia2_x,codigo_ocorrencia3,codigo_ocorrencia4,ocorrencia_classificacao,ocorrencia_latitude,ocorrencia_longitude,ocorrencia_cidade,ocorrencia_uf,...,aeronave_registro_segmento,aeronave_voo_origem,aeronave_voo_destino,aeronave_fase_operacao,aeronave_tipo_operacao,aeronave_nivel_dano,aeronave_fatalidades_total,codigo_ocorrencia1_y,ocorrencia_tipo_categoria,taxonomia_tipo_icao
0,87125,87125,87125,87125,87125,INCIDENTE,-7.219166666666,-39.26944444444,JUAZEIRO DO NORTE,CE,...,***,VIRACOPOS,ORLANDO BEZERRA DE MENEZES,DESCIDA,REGULAR,NENHUM,0.0,87125,FALHA OU MAU FUNCIONAMENTO DE SISTEMA / COMPON...,SCF-NP
1,87124,87124,87124,87124,87124,INCIDENTE,-18.88361111111,-48.22527777777,UBERLÂNDIA,MG,...,***,TENENTE-CORONEL AVIADOR CÉSAR BOMBONATO,TANCREDO NEVES,SUBIDA,REGULAR,NENHUM,0.0,87124,FALHA OU MAU FUNCIONAMENTO DE SISTEMA / COMPON...,SCF-NP
2,87123,87123,87123,87123,87123,INCIDENTE,-23.43555555555,-46.47305555555,GUARULHOS,SP,...,***,GOVERNADOR ANDRÉ FRANCO MONTORO,PINTO MARTINS,DECOLAGEM,REGULAR,NENHUM,0.0,87123,FALHA OU MAU FUNCIONAMENTO DE SISTEMA / COMPON...,SCF-NP


In [5]:
# Estatísticas das colunas numéricas
df.describe(include='all').T[['count', 'unique', 'top', 'freq', 'mean', 'std', 'min', 'max']]

,count,unique,top,freq,mean,std,min,max
codigo_ocorrencia,13301.0,NaN,NaN,NaN,67809.14638,18131.626733,28256.0,87125.0
codigo_ocorrencia1_x,13301.0,NaN,NaN,NaN,67809.14638,18131.626733,28256.0,87125.0
codigo_ocorrencia2_x,13301.0,NaN,NaN,NaN,67809.14638,18131.626733,28256.0,87125.0
codigo_ocorrencia3,13301.0,NaN,NaN,NaN,67809.14638,18131.626733,28256.0,87125.0
codigo_ocorrencia4,13301.0,NaN,NaN,NaN,67809.14638,18131.626733,28256.0,87125.0
ocorrencia_classificacao,13301,3,INCIDENTE,9252,NaN,NaN,NaN,NaN
ocorrencia_latitude,10525,3721,***,1283,NaN,NaN,NaN,NaN
ocorrencia_longitude,10525,3728,***,1284,NaN,NaN,NaN,NaN
ocorrencia_cidade,13301,1430,RIO DE JANEIRO,926,NaN,NaN,NaN,NaN
ocorrencia_uf,13301,28,SP,3323,NaN,NaN,NaN,NaN


### 3.2 — Análise de Valores Ausentes

O marcador `***` deve ser convertido para `NaN` antes de qualquer processamento. A proporção de ausentes varia entre variáveis: parte é estrutural — dados de investigação formal ausentes em ocorrências leves — e parte reflete falha de registro. A distinção é metodologicamente relevante: no primeiro caso, a ausência carrega informação implícita sobre a severidade do evento.

In [6]:
df.replace(['***', '', 'NULL'], np.nan, inplace=True)

missing = (df.isnull().sum() / len(df) * 100).sort_values(ascending=False)
missing_df = missing[missing > 0].reset_index()
missing_df.columns = ['coluna', 'pct_ausente']

fig = px.bar(
    missing_df.head(20),
    x='pct_ausente', y='coluna',
    orientation='h',
    title='Top 20 Colunas com Valores Ausentes (%)',
    labels={'pct_ausente': '% Ausente', 'coluna': ''},
    color='pct_ausente',
    color_continuous_scale='Reds'
)
fig.update_layout(yaxis=dict(autorange='reversed'))
fig.show()

print(f'\nColunas com >50% ausentes: {(missing > 50).sum()}')
print(f'Colunas sem ausentes:      {(missing == 0).sum()}')


Colunas com >50% ausentes: 3
Colunas sem ausentes:      17


### 3.3 — Distribuição da Variável-Alvo: Nível de Dano

Segue a taxonomia CENIPA/ICAO [1]: **NENHUM** (sem dano físico), **LEVE** (reparável), **SUBSTANCIAL** (dano estrutural extenso) e **DESTRUÍDA** (perda total). A distribuição empiricamente desbalanceada — DESTRUÍDA representa ~4% dos registros — constitui o principal desafio de modelagem.

O desbalanceamento de classes em domínios de segurança crítica é documentado na literatura [6]: eventos de maior severidade são intrinsecamente raros por resultarem do alinhamento simultâneo de múltiplas falhas em barreiras independentes — modelo de Reason [7]. A acurácia global é inadequada como métrica; o **F1-score ponderado** é adotado por penalizar modelos que ignoram classes minoritárias [8].

In [7]:
dano_counts = df['aeronave_nivel_dano'].value_counts(dropna=False)
print('Distribuição de aeronave_nivel_dano:')
print(dano_counts)

dano_df = dano_counts.reset_index()
dano_df.columns = ['nivel_dano', 'count']

fig = px.bar(
    dano_df, x='nivel_dano', y='count',
    title='Distribuição do Nível de Dano (Target)',
    labels={'nivel_dano': 'Nível de Dano', 'count': 'Quantidade'},
    color='count',
    color_continuous_scale='Oranges'
)
fig.show()

Distribuição de aeronave_nivel_dano:
aeronave_nivel_dano
NENHUM         6626
LEVE           3338
SUBSTANCIAL    2416
DESTRUÍDA       548
NaN             373
Name: count, dtype: int64


### 3.4 — Severidade por Tipo de Operação

A relação entre tipo de operação e severidade é examinada em termos de **proporção dentro de cada categoria**, não de contagem absoluta. Operações de baixa altitude (AGRÍCOLA) e aeronaves experimentais tendem a apresentar maior proporção de danos severos — padrão consistente com a literatura de fatores de risco em aviação geral [2].

**Nota sobre viés de exposição:** a ausência de denominador de exposição por segmento (horas voadas ou movimentos) impede o cálculo de taxas de risco absoluto. Os resultados descrevem severidade *dado que uma ocorrência foi registrada*, não risco por operação realizada.

In [8]:
cross = pd.crosstab(
    df['aeronave_tipo_operacao'],
    df['aeronave_nivel_dano'],
    normalize='index'
) * 100

if 'DESTRUÍDA' in cross.columns:
    cross = cross.sort_values('DESTRUÍDA', ascending=False)

cross_long = cross.reset_index().melt(
    id_vars='aeronave_tipo_operacao',
    var_name='nivel_dano',
    value_name='pct'
)

fig = px.bar(
    cross_long,
    x='aeronave_tipo_operacao', y='pct',
    color='nivel_dano',
    barmode='stack',
    title='Distribuição de Dano por Tipo de Operação (%)',
    labels={'aeronave_tipo_operacao': 'Tipo de Operação', 'pct': '%', 'nivel_dano': 'Nível de Dano'},
    color_discrete_sequence=px.colors.diverging.RdYlGn[::-1]
)
fig.update_layout(xaxis_tickangle=-45)
fig.show()

print('\n% por tipo de operação:')
cross.round(1)


% por tipo de operação:


aeronave_nivel_dano,DESTRUÍDA,LEVE,NENHUM,SUBSTANCIAL
aeronave_tipo_operacao,,,,
EXPERIMENTAL,17.1,24.0,14.6,44.2
AGRÍCOLA,11.1,17.2,4.3,67.3
PRIVADA,9.0,33.6,26.5,31.0
ESPECIALIZADA,5.8,28.6,39.0,26.6
POLICIAL,3.4,48.1,29.7,18.8
TÁXI AÉREO,3.3,32.6,48.7,15.3
INSTRUÇÃO,2.3,35.0,32.5,30.2
NÃO REGULAR,1.6,31.7,51.2,15.4
REGULAR,0.1,20.1,79.2,0.6


### 3.5 — Análise das Features Candidatas

Features com alto percentual de ausentes (> 30–40%), alta cardinalidade ou geradas como consequência da ocorrência são excluídas — as últimas por configurarem leakage (Seção 4.1).

A seleção foi orientada por dois critérios: **cobertura** (proporção de registros não nulos) e **precedência temporal** — a variável deve existir antes da ocorrência, evitando *data leakage* [4].

In [9]:
fase_counts = df['aeronave_fase_operacao'].value_counts().head(10).reset_index()
fase_counts.columns = ['fase', 'count']

motor_counts = df['aeronave_motor_tipo'].value_counts().head(8).reset_index()
motor_counts.columns = ['motor', 'count']

class_counts = df['ocorrencia_classificacao'].value_counts().reset_index()
class_counts.columns = ['classificacao', 'count']

fig1 = px.bar(
    fase_counts, x='count', y='fase', orientation='h',
    title='Top 10 Fases de Operação',
    labels={'count': 'Quantidade', 'fase': ''},
    color='count', color_continuous_scale='Blues'
)
fig1.update_layout(yaxis=dict(autorange='reversed'))
fig1.show()

fig2 = px.bar(
    motor_counts, x='count', y='motor', orientation='h',
    title='Tipo de Motor',
    labels={'count': 'Quantidade', 'motor': ''},
    color='count', color_continuous_scale='Greens'
)
fig2.update_layout(yaxis=dict(autorange='reversed'))
fig2.show()

fig3 = px.bar(
    class_counts, x='count', y='classificacao', orientation='h',
    title='Classificação da Ocorrência',
    labels={'count': 'Quantidade', 'classificacao': ''},
    color='count', color_continuous_scale='Oranges'
)
fig3.update_layout(yaxis=dict(autorange='reversed'))
fig3.show()

### 3.6 — Análise Temporal

A distribuição de severidade ao longo do tempo não é estacionária: a criação da ANAC em 2005 e a adoção dos padrões RBAC reduziram estruturalmente a frequência de eventos graves. Dados de décadas anteriores podem refletir uma realidade regulatória e tecnológica incompatível com o risco contemporâneo.

In [10]:
df['ocorrencia_dia'] = pd.to_datetime(df['ocorrencia_dia'], format='%d/%m/%Y', errors='coerce')
df['ano'] = df['ocorrencia_dia'].dt.year

por_ano = df.groupby('ano').size().reset_index(name='total')
por_ano = por_ano[por_ano['ano'].between(2010, 2025)]

fig = px.line(
    por_ano, x='ano', y='total',
    markers=True,
    title='Ocorrências por Ano (2010–2025)',
    labels={'ano': 'Ano', 'total': 'Total de Ocorrências'},
    color_discrete_sequence=['steelblue']
)
fig.update_xaxes(dtick=1)
fig.show()

---

## Seção 4 — Preparação dos Dados

Decisões de preparação têm implicações diretas sobre a confiabilidade das métricas e a capacidade de generalização dos modelos [4].

### 4.1 — Data Leakage: Definição e Controle

*Data leakage* ocorre quando informações indisponíveis no momento da predição são incorporadas ao treinamento, produzindo estimativas artificialmente otimistas [4]. Dois vetores foram identificados:

**Leakage por variável:** `fator_contribuinte` foi descartado. A taxa de registros com fator salta de 7,9% em LEVE para 57,0% em SUBSTANCIAL — mapeando o protocolo do CENIPA de abrir investigações em função da severidade. Usar essa variável incorporaria o target indiretamente ao modelo.

**Leakage por pré-processamento:** qualquer transformação que use estatísticas do dataset completo antes da divisão treino/teste contamina o conjunto de teste. Controlado pela ordem canônica: $\text{Split} \rightarrow \text{Fit} \rightarrow \text{Transform}$.

In [11]:
FEAT_CAT_P2 = [
    'aeronave_tipo_operacao',
    'ocorrencia_classificacao',
    'aeronave_fase_operacao',
    'aeronave_motor_tipo',
    'aeronave_motor_quantidade',
    'ocorrencia_tipo_categoria',
]
TARGET_P2 = 'aeronave_nivel_dano'

df_model = df[FEAT_CAT_P2 + [TARGET_P2]].copy()

print(f'Shape antes de limpeza: {df_model.shape}')
print(f'\nNaN por coluna:')
print(df_model.isnull().sum().sort_values(ascending=False))


Shape antes de limpeza: (13301, 7)

NaN por coluna:
aeronave_motor_tipo          1267
aeronave_tipo_operacao        989
aeronave_motor_quantidade     567
aeronave_nivel_dano           373
aeronave_fase_operacao         30
ocorrencia_tipo_categoria       7
ocorrencia_classificacao        0
dtype: int64


### 4.2 — Tratamento de Valores Ausentes

Valores ausentes categóricos são imputados pela **moda**, dentro do pipeline e após o split — preservando a distribuição sem criar categorias fora do suporte de treino.

In [12]:
# Remover linhas onde o TARGET é nulo (não há o que prever)
df_model = df_model.dropna(subset=[TARGET_P2])

# Remover classes muito raras (menos de 30 exemplos) — inviáveis para treino
min_samples = 30
class_counts = df_model[TARGET_P2].value_counts()
classes_validas = class_counts[class_counts >= min_samples].index
df_model = df_model[df_model[TARGET_P2].isin(classes_validas)]

df_p2 = df_model  # alias usado na modelagem por tipo de operação

print(f'Shape após limpeza: {df_model.shape}')
print(f'\nClasses mantidas ({len(classes_validas)}):')
print(df_model[TARGET_P2].value_counts())


Shape após limpeza: (12928, 7)

Classes mantidas (4):
aeronave_nivel_dano
NENHUM         6626
LEVE           3338
SUBSTANCIAL    2416
DESTRUÍDA       548
Name: count, dtype: int64


### 4.3 — Encoding de Variáveis Categóricas

**OneHotEncoding** é adotado para features categóricas nominais: evita a ordinalidade artificial do *label encoding*, que introduziria relações de ordem sem fundamento no domínio [9]. Para o target, `LabelEncoder` produz os rótulos inteiros esperados pelos classificadores.

In [13]:
X = df_model[FEAT_CAT_P2].copy()
y = df_model[TARGET_P2].copy()

print(f'X: {X.shape} | y: {y.shape}')
print(f'Classes: {sorted(y.unique())}')


X: (12928, 6) | y: (12928,)
Classes: ['DESTRUÍDA', 'LEVE', 'NENHUM', 'SUBSTANCIAL']


### 4.4 — Divisão Treino/Teste e Pipeline

A divisão 80/20 é realizada **antes** de qualquer pré-processamento, com `stratify=y` para preservar a proporção de classes em datasets desbalanceados. O `Pipeline` do scikit-learn encapsula imputação e codificação, garantindo que o conjunto de teste nunca influencie os parâmetros aprendidos no treino. Classes com suporte < 30 são removidas antes do split — abaixo desse limiar, métricas por classe são estatisticamente não confiáveis [8].

In [14]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=RANDOM_STATE,
    stratify=y
)

print(f'Treino: {X_train.shape[0]} amostras | Teste: {X_test.shape[0]} amostras')
print(f'Proporção treino/teste: {X_train.shape[0]/len(X)*100:.0f}% / {X_test.shape[0]/len(X)*100:.0f}%')
print()
print('Distribuição de classes no treino:')
print(y_train.value_counts(normalize=True).round(3))

Treino: 10342 amostras | Teste: 2586 amostras
Proporção treino/teste: 80% / 20%

Distribuição de classes no treino:
aeronave_nivel_dano
NENHUM         0.513
LEVE           0.258
SUBSTANCIAL    0.187
DESTRUÍDA      0.042
Name: proportion, dtype: float64


### 4.5 — Pipeline de Pré-processamento

O conceito de pipeline aqui resolve um problema elegante: como garantir que o pré-processamento do conjunto de teste use exatamente os mesmos parâmetros aprendidos no treino?

A resposta é encapsular a imputação e o encoding num objeto `Pipeline` do scikit-learn. Esse objeto é ajustado (`fit`) **apenas nos dados de treino** — aprende a moda de cada coluna, aprende as categorias válidas do OHE. Quando aplicado (`transform`) ao conjunto de teste, usa os mesmos parâmetros sem "olhar" para os dados de teste.

Se o conjunto de teste tiver uma categoria nova que não apareceu no treino, o parâmetro `handle_unknown='ignore'` garante que ela seja simplesmente ignorada, em vez de causar um erro.

In [15]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

cat_transformer_p2 = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ohe', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

preprocessor_p2 = ColumnTransformer(transformers=[
    ('cat', cat_transformer_p2, FEAT_CAT_P2)
])

X_train_p2_prep = preprocessor_p2.fit_transform(X_train)
X_test_p2_prep  = preprocessor_p2.transform(X_test)
X_test_p2_raw   = X_test.copy()

print(f'X_train pré-processado: {X_train_p2_prep.shape}')
print(f'X_test  pré-processado: {X_test_p2_prep.shape}')


X_train pré-processado: (10342, 139)


X_test  pré-processado: (2586, 139)


### 4.6 — Encoding do Target

O target também precisa ser codificado — mas aqui usamos `LabelEncoder`, não OHE. O motivo é diferente: o target não é uma feature de entrada, é a saída que o modelo deve prever. Os algoritmos de classificação esperam rótulos inteiros (0, 1, 2, 3), não colunas binárias. O `LabelEncoder` faz exatamente essa conversão, mapeando DESTRUÍDA, LEVE, NENHUM, SUBSTANCIAL para 0, 1, 2, 3 em ordem alfabética.


In [16]:
le_p2 = LabelEncoder()
y_train_p2_enc = le_p2.fit_transform(y_train)
y_test_p2_enc  = le_p2.transform(y_test)

print('Mapeamento de classes:')
for i, cls in enumerate(le_p2.classes_):
    print(f'  {i} → {cls}')


Mapeamento de classes:
  0 → DESTRUÍDA
  1 → LEVE
  2 → NENHUM
  3 → SUBSTANCIAL


### 4.7 — Resumo da Preparação

| Etapa | Decisão | Justificativa |
|---|---|---|
| Marcador `***` | Conversão para `NaN` | Permite imputação nativa |
| Missing categórico | Moda, no pipeline, após split | Preserva distribuição; evita leakage |
| Encoding de features | OneHotEncoding | Evita ordinalidade artificial |
| Encoding do target | LabelEncoding | Formato esperado pelo classificador |
| Classes raras (< 30) | Remoção antes do split | Métricas por classe não confiáveis |
| Divisão | 80/20 estratificado, pré-processamento posterior | Previne leakage |
| `fator_contribuinte` | Descartado | Leakage por variável |

---

## Seção 5 — Modelagem: Previsão de Severidade

### 5.1 Estratégias: Opção A e Opção B

| | Opção A | Opção B |
|---|---|---|
| **Estratégia** | XGBoost dedicado por tipo de operação | Modelo único com tipo de operação como feature OHE |
| **Hipótese** | Padrões de severidade são suficientemente distintos entre tipos para justificar especialização | O modelo único aprende as interações implicitamente |
| **Vantagem** | Captura padrões específicos de cada contexto | Estável para tipos com poucos dados; um único modelo |
| **Risco** | Tipos raros produzem modelos frágeis | Pode suavizar diferenças reais entre tipos |

Ambas as hipóteses são testadas empiricamente; o benchmark é apresentado na Seção 6.3.


### 5.2 — Preparação e Implementação: Opção B

O target é `aeronave_nivel_dano`; registros nulos nessa variável são removidos; `aeronave_tipo_operacao` entra no pipeline OHE como feature. Uma cópia do conjunto de teste com valores originais (`X_test_p2_raw`) é preservada para o benchmark por tipo de operação.

Na **Opção B**, quatro modelos são treinados sobre o conjunto completo (~12.900 registros):

- **Baseline (DummyClassifier, `most_frequent`):** sempre prevê a classe majoritária (NENHUM), ignorando as features. Serve como piso de referência — qualquer modelo real precisa superá-lo para justificar o uso de Machine Learning.
- **Logistic Regression** e **Random Forest**: dois candidatos de famílias distintas (linear e ensemble de árvores), usados para avaliar se a relação entre features e severidade é predominantemente linear ou beneficia de interações não lineares.
- **XGBoost**: candidato principal, escolhido por lidar bem com features categóricas esparsas (pós-OHE) e classes desbalanceadas.


In [17]:
from sklearn.dummy import DummyClassifier

# Baseline simples: sempre prevê a classe majoritária, ignorando as features
baseline_p2 = DummyClassifier(strategy='most_frequent', random_state=RANDOM_STATE)
baseline_p2.fit(X_train_p2_prep, y_train_p2_enc)
y_pred_baseline_p2 = baseline_p2.predict(X_test_p2_prep)

lr_p2 = LogisticRegression(max_iter=1000, random_state=RANDOM_STATE)
lr_p2.fit(X_train_p2_prep, y_train_p2_enc)
y_pred_lr_p2 = lr_p2.predict(X_test_p2_prep)

rf_p2 = RandomForestClassifier(n_estimators=100, random_state=RANDOM_STATE, n_jobs=-1)
rf_p2.fit(X_train_p2_prep, y_train_p2_enc)
y_pred_rf_p2 = rf_p2.predict(X_test_p2_prep)

xgb_p2 = XGBClassifier(random_state=RANDOM_STATE, eval_metric='mlogloss', verbosity=0)
xgb_p2.fit(X_train_p2_prep, y_train_p2_enc)
y_pred_xgb_p2 = xgb_p2.predict(X_test_p2_prep)

print('=== Opção B — Resultados ===')
for name, y_pred in [
    ('Baseline (classe majoritária)', y_pred_baseline_p2),
    ('Logistic Regression',           y_pred_lr_p2),
    ('Random Forest',                 y_pred_rf_p2),
    ('XGBoost',                       y_pred_xgb_p2),
]:
    acc = accuracy_score(y_test_p2_enc, y_pred)
    f1  = f1_score(y_test_p2_enc, y_pred, average='weighted', zero_division=0)
    print(f'{name:<32} Acc={acc:.4f} | F1={f1:.4f}')


=== Opção B — Resultados ===
Baseline (classe majoritária)    Acc=0.5124 | F1=0.3472
Logistic Regression              Acc=0.7657 | F1=0.7587
Random Forest                    Acc=0.7633 | F1=0.7610
XGBoost                          Acc=0.7776 | F1=0.7747


### 5.3 — Opção A: XGBoost por Tipo de Operação

Na **Opção A**, um XGBoost independente é treinado por segmento. Tipos com menos de 100 registros são excluídos: o volume precisa sustentar até quatro classes de severidade após a divisão treino/teste — exigência mais restritiva que o limiar de 30 da Seção 4.4.


In [18]:
FEAT_OPTA = [f for f in FEAT_CAT_P2 if f != 'aeronave_tipo_operacao']
MIN_SAMPLES_TYPE = 100
opta_results = []
opta_models = {}

op_types = sorted(df_p2['aeronave_tipo_operacao'].dropna().unique())

for op_type in op_types:
    subset = df_p2[df_p2['aeronave_tipo_operacao'] == op_type].copy()

    if len(subset) < MIN_SAMPLES_TYPE:
        print(f'[SKIP] {op_type}: {len(subset)} amostras < {MIN_SAMPLES_TYPE}')
        continue

    X_sub = subset[FEAT_OPTA]
    y_sub = subset[TARGET_P2]

    # Drop severity classes with < 5 samples to allow stratified split
    valid_cls = y_sub.value_counts()[y_sub.value_counts() >= 5].index
    X_sub = X_sub[y_sub.isin(valid_cls)]
    y_sub = y_sub[y_sub.isin(valid_cls)]

    if y_sub.nunique() < 2:
        print(f'[SKIP] {op_type}: apenas 1 classe de severidade com dados suficientes')
        continue

    try:
        X_tr, X_te, y_tr, y_te = train_test_split(
            X_sub, y_sub, test_size=0.2, stratify=y_sub, random_state=RANDOM_STATE
        )
    except ValueError:
        X_tr, X_te, y_tr, y_te = train_test_split(
            X_sub, y_sub, test_size=0.2, random_state=RANDOM_STATE
        )

    cat_sub = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('ohe', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
    ])
    prep_sub = ColumnTransformer(transformers=[('cat', cat_sub, FEAT_OPTA)])
    X_tr_p = prep_sub.fit_transform(X_tr)
    X_te_p = prep_sub.transform(X_te)

    le_sub = LabelEncoder()
    y_tr_e = le_sub.fit_transform(y_tr)
    y_te_e = le_sub.transform(y_te)

    m = XGBClassifier(random_state=RANDOM_STATE, eval_metric='mlogloss', verbosity=0)
    m.fit(X_tr_p, y_tr_e)
    y_pred_sub = m.predict(X_te_p)

    f1  = f1_score(y_te_e, y_pred_sub, average='weighted', zero_division=0)
    acc = accuracy_score(y_te_e, y_pred_sub)

    opta_results.append({
        'Tipo de Operação': op_type,
        'Amostras': len(subset),
        'Classes Severidade': y_sub.nunique(),
        'Acurácia': round(acc, 4),
        'F1 weighted': round(f1, 4),
    })
    opta_models[op_type] = {'preprocessor': prep_sub, 'le': le_sub, 'model': m}
    print(f'{op_type:<18} F1={f1:.4f} | Acc={acc:.4f} | n={len(subset)} | classes={y_sub.nunique()}')

opta_df = pd.DataFrame(opta_results).sort_values('F1 weighted', ascending=False)
print(f'\n{len(opta_df)} modelos treinados com sucesso.')

[SKIP] AERODESPORTIVA: 15 amostras < 100


AGRÍCOLA           F1=0.7707 | Acc=0.8012 | n=830 | classes=4


ESPECIALIZADA      F1=0.5606 | Acc=0.5484 | n=154 | classes=4


EXPERIMENTAL       F1=0.6324 | Acc=0.6301 | n=362 | classes=4


INSTRUÇÃO          F1=0.7138 | Acc=0.7149 | n=1137 | classes=4


NÃO REGULAR        F1=0.4405 | Acc=0.4400 | n=123 | classes=3


POLICIAL           F1=0.5683 | Acc=0.5926 | n=266 | classes=4


PRIVADA            F1=0.7182 | Acc=0.7236 | n=3129 | classes=4


REGULAR            F1=0.8078 | Acc=0.8342 | n=4619 | classes=3


TÁXI AÉREO         F1=0.7288 | Acc=0.7311 | n=1316 | classes=4

9 modelos treinados com sucesso.


---

## Seção 6 — Avaliação

Com os modelos treinados, avaliamos quatro conjuntos de resultados: o baseline, o desempenho geral da Opção B, o desempenho por tipo da Opção A, e o benchmark direto entre as abordagens. A atenção especial recai sobre DESTRUÍDA: falsos negativos têm consequências de segurança mais graves que falsos positivos.


### 6.1 — Avaliação da Opção B (Baseline e Modelos Únicos)

O baseline (classe majoritária) fixa o piso de comparação: qualquer ganho de F1 dos demais modelos sobre ele é a evidência de que as features carregam sinal preditivo real. A atenção recai sobre DESTRUÍDA: falsos negativos — predizer LEVE quando o dano real é DESTRUÍDA — têm consequências de segurança mais graves que falsos positivos. Essa assimetria sugere calibração de limiar de decisão em aplicações operacionais futuras.


In [19]:
from sklearn.metrics import precision_score, recall_score

results_p2 = pd.DataFrame({
    'Modelo': ['Baseline (classe majoritária)', 'Logistic Regression', 'Random Forest', 'XGBoost'],
    'Acurácia': [
        accuracy_score(y_test_p2_enc, y_pred_baseline_p2),
        accuracy_score(y_test_p2_enc, y_pred_lr_p2),
        accuracy_score(y_test_p2_enc, y_pred_rf_p2),
        accuracy_score(y_test_p2_enc, y_pred_xgb_p2),
    ],
    'Precisão (w)': [
        precision_score(y_test_p2_enc, y_pred_baseline_p2, average='weighted', zero_division=0),
        precision_score(y_test_p2_enc, y_pred_lr_p2, average='weighted', zero_division=0),
        precision_score(y_test_p2_enc, y_pred_rf_p2, average='weighted', zero_division=0),
        precision_score(y_test_p2_enc, y_pred_xgb_p2, average='weighted', zero_division=0),
    ],
    'Recall (w)': [
        recall_score(y_test_p2_enc, y_pred_baseline_p2, average='weighted', zero_division=0),
        recall_score(y_test_p2_enc, y_pred_lr_p2, average='weighted', zero_division=0),
        recall_score(y_test_p2_enc, y_pred_rf_p2, average='weighted', zero_division=0),
        recall_score(y_test_p2_enc, y_pred_xgb_p2, average='weighted', zero_division=0),
    ],
    'F1 (w)': [
        f1_score(y_test_p2_enc, y_pred_baseline_p2, average='weighted', zero_division=0),
        f1_score(y_test_p2_enc, y_pred_lr_p2, average='weighted', zero_division=0),
        f1_score(y_test_p2_enc, y_pred_rf_p2, average='weighted', zero_division=0),
        f1_score(y_test_p2_enc, y_pred_xgb_p2, average='weighted', zero_division=0),
    ],
})

print(results_p2.round(4).to_string(index=False))

fig = px.bar(
    results_p2.melt(id_vars='Modelo', var_name='Métrica', value_name='Valor'),
    x='Modelo', y='Valor', color='Métrica', barmode='group',
    title='Opção B — Baseline vs. Modelos Candidatos (Severidade)',
    labels={'Valor': 'Score', 'Modelo': ''},
    text_auto='.3f'
)
fig.update_layout(yaxis_range=[0, 1])
fig.show()

cm_p2 = confusion_matrix(y_test_p2_enc, y_pred_xgb_p2)
fig_cm = px.imshow(
    cm_p2,
    labels=dict(x='Predito', y='Real', color='Contagem'),
    x=list(le_p2.classes_),
    y=list(le_p2.classes_),
    text_auto=True,
    color_continuous_scale='Blues',
    title='Matriz de Confusão — XGBoost (Opção B)'
)
fig_cm.show()


                       Modelo  Acurácia  Precisão (w)  Recall (w)  F1 (w)
Baseline (classe majoritária)    0.5124        0.2625      0.5124  0.3472
          Logistic Regression    0.7657        0.7557      0.7657  0.7587
                Random Forest    0.7633        0.7592      0.7633  0.7610
                      XGBoost    0.7776        0.7727      0.7776  0.7747


In [20]:
print('=== Classification Report — XGBoost (Opção B) ===')
print(classification_report(y_test_p2_enc, y_pred_xgb_p2, target_names=le_p2.classes_))

=== Classification Report — XGBoost (Opção B) ===
              precision    recall  f1-score   support

   DESTRUÍDA       0.49      0.38      0.43       110
        LEVE       0.67      0.64      0.65       668
      NENHUM       0.84      0.87      0.85      1325
 SUBSTANCIAL       0.80      0.80      0.80       483

    accuracy                           0.78      2586
   macro avg       0.70      0.67      0.68      2586
weighted avg       0.77      0.78      0.77      2586



### 6.2 — Avaliação da Opção A (Modelo por Tipo de Operação)

Produz um perfil de F1 por segmento, não uma métrica global. O padrão esperado é monotônico com o volume: tipos com mais registros produzem modelos mais estáveis; tipos raros apresentam alta variância independentemente do algoritmo [6].


In [21]:
print('=== Sumário Opção A — XGBoost por Tipo de Operação ===')
print(opta_df.to_string(index=False))

fig = px.bar(
    opta_df.sort_values('F1 weighted', ascending=True),
    x='F1 weighted', y='Tipo de Operação',
    orientation='h',
    title='Opção A — F1 Weighted por Tipo de Operação (XGBoost dedicado)',
    labels={'F1 weighted': 'F1 (weighted)', 'Tipo de Operação': ''},
    color='F1 weighted',
    color_continuous_scale='Blues',
    text_auto='.3f'
)
fig.update_layout(xaxis_range=[0, 1])
fig.show()

=== Sumário Opção A — XGBoost por Tipo de Operação ===
Tipo de Operação  Amostras  Classes Severidade  Acurácia  F1 weighted
         REGULAR      4619                   3    0.8342       0.8078
        AGRÍCOLA       830                   4    0.8012       0.7707
      TÁXI AÉREO      1316                   4    0.7311       0.7288
         PRIVADA      3129                   4    0.7236       0.7182
       INSTRUÇÃO      1137                   4    0.7149       0.7138
    EXPERIMENTAL       362                   4    0.6301       0.6324
        POLICIAL       266                   4    0.5926       0.5683
   ESPECIALIZADA       154                   4    0.5484       0.5606
     NÃO REGULAR       123                   3    0.4400       0.4405


### 6.3 — Benchmark: Opção A vs. Opção B

$$\Delta F1 = F1_{\text{Opção A}} - F1_{\text{Opção B}}$$

Delta positivo indica que a especialização superou o modelo único para aquele tipo. Delta negativo indica o contrário. Diferenças < 0,02 estão dentro da variância típica de F1 nessa escala e não constituem evidência de superioridade [8]. A hipótese da Opção A é sustentada apenas onde Δ é positivo e de magnitude relevante (> 0,10).


In [22]:
benchmark = []

for _, row in opta_df.iterrows():
    op_type = row['Tipo de Operação']
    f1_a    = row['F1 weighted']

    mask = X_test_p2_raw['aeronave_tipo_operacao'] == op_type
    if mask.sum() < 5:
        continue

    X_type_prep = preprocessor_p2.transform(X_test_p2_raw[mask])
    y_true_enc  = le_p2.transform(y_test[mask])
    y_pred_type = xgb_p2.predict(X_type_prep)

    f1_b = f1_score(y_true_enc, y_pred_type, average='weighted', zero_division=0)

    benchmark.append({
        'Tipo de Operação':  op_type,
        'F1 Opção A':        round(f1_a, 4),
        'F1 Opção B':        round(f1_b, 4),
        'Delta (A − B)':     round(f1_a - f1_b, 4),
    })

bench_df = pd.DataFrame(benchmark).sort_values('Delta (A − B)', ascending=False)
print(bench_df.to_string(index=False))

bench_long = bench_df.melt(
    id_vars='Tipo de Operação',
    value_vars=['F1 Opção A', 'F1 Opção B'],
    var_name='Abordagem',
    value_name='F1 weighted'
)

fig = px.bar(
    bench_long,
    x='F1 weighted',
    y='Tipo de Operação',
    color='Abordagem',
    barmode='group',
    orientation='h',
    title='Benchmark: Opção A vs Opção B — F1 por Tipo de Operação',
    labels={'F1 weighted': 'F1 (weighted)', 'Tipo de Operação': ''},
    text_auto='.3f'
)
fig.update_layout(xaxis_range=[0, 1])
fig.show()

Tipo de Operação  F1 Opção A  F1 Opção B  Delta (A − B)
        AGRÍCOLA      0.7707      0.7476         0.0231
      TÁXI AÉREO      0.7288      0.7236         0.0052
         REGULAR      0.8078      0.8063         0.0015
       INSTRUÇÃO      0.7138      0.7136         0.0002
         PRIVADA      0.7182      0.7195        -0.0013
   ESPECIALIZADA      0.5606      0.6349        -0.0743
    EXPERIMENTAL      0.6324      0.7342        -0.1018
        POLICIAL      0.5683      0.7381        -0.1698
     NÃO REGULAR      0.4405      0.6277        -0.1872


### 6.4 — Otimização de Hiperparâmetros

Hiperparâmetros controlam *como* o algoritmo aprende e afetam diretamente o tradeoff viés-variância [10]. A busca sistemática é etapa metodológica necessária, não refinamento opcional.

**RandomizedSearchCV:** Bergstra e Bengio [15] demonstraram que a busca aleatória converge para boas regiões do espaço de hiperparâmetros mais eficientemente que o *grid search* exaustivo quando apenas um subconjunto de parâmetros é determinante. Com `n_iter=30` e validação cruzada de 5 folds, são realizadas 150 avaliações — aplicadas ao **XGBoost da Opção B**, vencedor do benchmark da Seção 6.3.

**Parâmetros e Justificativas:**

| Parâmetro | Função | Faixa |
|---|---|---|
| `n_estimators` | Número de árvores sequenciais | 100–500 |
| `max_depth` | Profundidade máxima por árvore | 3–8 |
| `learning_rate` | Fator de encolhimento por iteração | 0,01–0,30 |
| `subsample` | Fração de instâncias por árvore | 0,6–1,0 |
| `colsample_bytree` | Fração de features por árvore | 0,6–1,0 |
| `min_child_weight` | Peso mínimo em nó folha | 1–7 |
| `gamma` | Ganho mínimo para realizar um split | 0–0,5 |

`learning_rate` e `n_estimators` são interdependentes: taxas menores exigem mais árvores, mas generalizam melhor. `subsample` e `colsample_bytree` introduzem regularização estocástica. `gamma` e `min_child_weight` são especialmente relevantes para DESTRUÍDA: com ~4% de representação, nós folha dessa classe têm poucos exemplos e parâmetros mal calibrados podem suprimi-los.

A busca é realizada exclusivamente sobre os dados de treino, com o conjunto de teste isolado até a avaliação final — prevenindo a seleção otimista de hiperparâmetros [4].


In [23]:
from sklearn.model_selection import RandomizedSearchCV

param_dist = {
    'n_estimators':      [100, 200, 300, 400, 500],
    'max_depth':         [3, 4, 5, 6, 8],
    'learning_rate':     [0.01, 0.05, 0.1, 0.2, 0.3],
    'subsample':         [0.6, 0.7, 0.8, 0.9, 1.0],
    'colsample_bytree':  [0.6, 0.7, 0.8, 0.9, 1.0],
    'min_child_weight':  [1, 3, 5, 7],
    'gamma':             [0, 0.1, 0.2, 0.3, 0.5],
}

xgb_base_tuning = XGBClassifier(
    random_state=RANDOM_STATE,
    eval_metric='mlogloss',
    verbosity=0
)

rs = RandomizedSearchCV(
    estimator=xgb_base_tuning,
    param_distributions=param_dist,
    n_iter=30,
    scoring='f1_weighted',
    cv=5,
    n_jobs=-1,
    random_state=RANDOM_STATE,
    verbose=1
)

print('Iniciando busca — 30 iterações × 5 folds = 150 avaliações...')
rs.fit(X_train_p2_prep, y_train_p2_enc)

print(f'\nMelhor F1 (CV): {rs.best_score_:.4f}')
print(f'\nMelhores hiperparâmetros:')
for k, v in rs.best_params_.items():
    print(f'  {k}: {v}')

Iniciando busca — 30 iterações × 5 folds = 150 avaliações...
Fitting 5 folds for each of 30 candidates, totalling 150 fits



Melhor F1 (CV): 0.7800

Melhores hiperparâmetros:
  subsample: 0.9
  n_estimators: 400
  min_child_weight: 7
  max_depth: 4
  learning_rate: 0.1
  gamma: 0.1
  colsample_bytree: 0.8


### Avaliação do Modelo Tunado

Agora usamos o melhor conjunto de parâmetros encontrado para avaliar no conjunto de teste — que não foi visto em nenhum momento durante a busca. O objetivo é comparar diretamente com o XGBoost padrão para quantificar o ganho real da otimização.

In [24]:
xgb_tuned = rs.best_estimator_
y_pred_tuned = xgb_tuned.predict(X_test_p2_prep)

# Métricas: tunado vs. baseline
acc_base  = accuracy_score(y_test_p2_enc, y_pred_xgb_p2)
f1_base   = f1_score(y_test_p2_enc, y_pred_xgb_p2,  average='weighted', zero_division=0)
acc_tuned = accuracy_score(y_test_p2_enc, y_pred_tuned)
f1_tuned  = f1_score(y_test_p2_enc, y_pred_tuned, average='weighted', zero_division=0)

# Destruída recall: o indicador mais crítico
from sklearn.metrics import recall_score
dest_idx = list(le_p2.classes_).index('DESTRUÍDA')

rec_base_dest  = recall_score(y_test_p2_enc, y_pred_xgb_p2,  average=None, zero_division=0)[dest_idx]
rec_tuned_dest = recall_score(y_test_p2_enc, y_pred_tuned, average=None, zero_division=0)[dest_idx]

comparison = pd.DataFrame({
    'Modelo':            ['XGBoost Padrão', 'XGBoost Tunado'],
    'Acurácia':          [round(acc_base, 4),  round(acc_tuned, 4)],
    'F1 weighted':       [round(f1_base, 4),   round(f1_tuned, 4)],
    'Recall DESTRUÍDA':  [round(rec_base_dest, 4), round(rec_tuned_dest, 4)],
    'Delta F1':          ['-', f'{f1_tuned - f1_base:+.4f}'],
    'Delta Recall DEST': ['-', f'{rec_tuned_dest - rec_base_dest:+.4f}'],
})

print('=== Comparação: Padrão vs Tunado ===')
print(comparison.to_string(index=False))

fig = px.bar(
    comparison[['Modelo', 'Acurácia', 'F1 weighted', 'Recall DESTRUÍDA']].melt(
        id_vars='Modelo', var_name='Métrica', value_name='Valor'
    ),
    x='Modelo', y='Valor', color='Métrica', barmode='group',
    title='XGBoost Padrão vs Tunado — Métricas Comparadas',
    labels={'Valor': 'Score', 'Modelo': ''},
    text_auto='.4f'
)
fig.update_layout(yaxis_range=[0, 1])
fig.show()

=== Comparação: Padrão vs Tunado ===
        Modelo  Acurácia  F1 weighted  Recall DESTRUÍDA Delta F1 Delta Recall DEST
XGBoost Padrão    0.7776       0.7747            0.3818        -                 -
XGBoost Tunado    0.7734       0.7677            0.2636  -0.0070           -0.1182


### Matriz de Confusão — XGBoost Tunado

A matriz de confusão do modelo tunado revela se a otimização alterou os padrões de erro — especialmente para a classe DESTRUÍDA, onde um ganho de recall, mesmo que pequeno, representa um impacto real na capacidade do modelo de identificar os acidentes mais graves.

In [25]:
cm_tuned = confusion_matrix(y_test_p2_enc, y_pred_tuned)

fig = px.imshow(
    cm_tuned,
    labels=dict(x='Predito', y='Real', color='Contagem'),
    x=list(le_p2.classes_),
    y=list(le_p2.classes_),
    text_auto=True,
    color_continuous_scale='Blues',
    title='Matriz de Confusão — XGBoost Tunado'
)
fig.show()

In [26]:
print('=== Classification Report — XGBoost Tunado ===')
print(classification_report(y_test_p2_enc, y_pred_tuned, target_names=le_p2.classes_))

print('\n=== Parâmetros do Melhor Modelo ===')
for k, v in rs.best_params_.items():
    print(f'  {k}: {v}')

=== Classification Report — XGBoost Tunado ===
              precision    recall  f1-score   support

   DESTRUÍDA       0.42      0.26      0.32       110
        LEVE       0.68      0.63      0.65       668
      NENHUM       0.84      0.87      0.86      1325
 SUBSTANCIAL       0.77      0.81      0.79       483

    accuracy                           0.77      2586
   macro avg       0.68      0.64      0.65      2586
weighted avg       0.76      0.77      0.77      2586


=== Parâmetros do Melhor Modelo ===
  subsample: 0.9
  n_estimators: 400
  min_child_weight: 7
  max_depth: 4
  learning_rate: 0.1
  gamma: 0.1
  colsample_bytree: 0.8


### 6.5 — Importância de Features (XGBoost Tunado)

A importância de features do modelo tunado revela quais variáveis o XGBoost usou mais para separar as classes de severidade. `aeronave_tipo_operacao` é uma das features — e a pergunta é se ela é dominante (o que poderia indicar que o modelo aprende principalmente pelo tipo) ou se o aprendizado é distribuído entre as features disponíveis.

Nota: importância de Gini pode superestimar variáveis de alta cardinalidade [14].


In [27]:
feat_names_p2 = preprocessor_p2.get_feature_names_out()
importances_p2 = xgb_tuned.feature_importances_

imp_df_p2 = (
    pd.DataFrame({'Feature': feat_names_p2, 'Importância': importances_p2})
    .sort_values('Importância', ascending=False)
    .head(20)
)

fig = px.bar(
    imp_df_p2.sort_values('Importância', ascending=True),
    x='Importância', y='Feature',
    orientation='h',
    title='Top 20 Features por Importância — XGBoost Tunado',
    labels={'Importância': 'Importância', 'Feature': ''},
    color='Importância',
    color_continuous_scale='Blues',
    text_auto='.3f'
)
fig.update_layout(showlegend=False)
fig.show()

print('\nTop 10 features:')
print(imp_df_p2[['Feature','Importância']].head(10).to_string(index=False))


Top 10 features:
                                                                                                                    Feature  Importância
                                                                                     cat__ocorrencia_classificacao_ACIDENTE     0.213866
                                                                                    cat__ocorrencia_classificacao_INCIDENTE     0.135908
                                          cat__ocorrencia_tipo_categoria_FALHA OU MAU FUNCIONAMENTO DE SISTEMA / COMPONENTE     0.042315
                                                                                        cat__aeronave_tipo_operacao_REGULAR     0.037197
                                                                               cat__aeronave_fase_operacao_REVISÃO DE PISTA     0.035564
                                         cat__ocorrencia_tipo_categoria_PERDA DE SEPARAÇÃO / COLISÃO EM VOO | TRÁFEGO AÉREO     0.034237
                       

### 6.6 — Diagnóstico de Generalização

O diagnóstico compara:

$$F1_{\text{treino}} \quad \longleftrightarrow \quad F1_{\text{CV}} \quad \longleftrightarrow \quad F1_{\text{teste}}$$

**Padrão saudável:** $F1_{\text{treino}} > F1_{\text{CV}} \approx F1_{\text{teste}}$, gap < 0,05. **Overfitting:** gap treino–teste > 0,08 — resposta: aumentar regularização. **Underfitting:** ambos baixos — resposta: revisar features ou aumentar complexidade.

O ganho incremental esperado da otimização é modesto: com seis variáveis categóricas, o modelo padrão já opera próximo do teto de informação disponível.


In [28]:
# F1 no treino — o modelo "decorou"?
y_train_pred_tuned = xgb_tuned.predict(X_train_p2_prep)
f1_train = f1_score(y_train_p2_enc, y_train_pred_tuned, average='weighted', zero_division=0)

# F1 no teste — o modelo generaliza?
f1_test  = f1_score(y_test_p2_enc, y_pred_tuned, average='weighted', zero_division=0)

# F1 na validação cruzada (melhor estimativa de performance real)
f1_cv    = rs.best_score_

gap = f1_train - f1_test

print('=== Diagnóstico de Overfitting / Underfitting ===')
print(f'  F1 Treino (XGBoost tunado):          {f1_train:.4f}')
print(f'  F1 Teste  (XGBoost tunado):          {f1_test:.4f}')
print(f'  F1 CV     (melhor config, 5-fold):   {f1_cv:.4f}')
print(f'  Gap Treino−Teste:                    {gap:+.4f}')
print()

if gap < 0.02:
    diagnostico = 'Saudável — sem sinais de overfitting. O modelo generaliza bem.'
elif gap < 0.08:
    diagnostico = 'Overfitting leve — o modelo aprendeu algum ruído do treino, mas a generalização é aceitável.'
else:
    diagnostico = 'Overfitting significativo — considere regularização adicional (menor max_depth, maior gamma).'

print(f'  Diagnóstico: {diagnostico}')

# Visual
diag_df = pd.DataFrame({
    'Conjunto': ['Treino', 'Validação Cruzada (CV)', 'Teste'],
    'F1 weighted': [f1_train, f1_cv, f1_test]
})

fig = px.bar(
    diag_df,
    x='Conjunto', y='F1 weighted',
    title='F1 Weighted — Treino vs CV vs Teste (XGBoost Tunado)',
    labels={'F1 weighted': 'F1 (weighted)', 'Conjunto': ''},
    color='Conjunto',
    color_discrete_sequence=px.colors.qualitative.Set2,
    text_auto='.4f'
)
fig.update_layout(yaxis_range=[0, 1], showlegend=False)
fig.show()

=== Diagnóstico de Overfitting / Underfitting ===
  F1 Treino (XGBoost tunado):          0.8005
  F1 Teste  (XGBoost tunado):          0.7677
  F1 CV     (melhor config, 5-fold):   0.7800
  Gap Treino−Teste:                    +0.0328

  Diagnóstico: Overfitting leve — o modelo aprendeu algum ruído do treino, mas a generalização é aceitável.


#### Interpretação

O gráfico conta a história em três pontos:

- **Treino > CV ≈ Teste** com gap pequeno (< 0.02): comportamento esperado e saudável. O modelo aprende mais do treino do que verá no mundo real, mas a diferença é marginal — não há memorização prejudicial.
- **CV próximo do Teste**: confirma que a validação cruzada foi um estimador confiável do desempenho real. O resultado do teste não foi sorte de um split favorável.
- **Treino muito acima de Teste** (gap > 0.08): sinalizaria overfitting. Nesse caso, os próximos passos seriam reduzir `max_depth`, aumentar `gamma`, ou reduzir `n_estimators`.

O XGBoost, por ser um modelo de gradient boosting com regularização nativa (`gamma`, `min_child_weight`), tende a ser menos propenso a overfitting severo do que uma árvore de decisão isolada — especialmente quando `subsample` e `colsample_bytree` estão abaixo de 1.0, como explorado na busca de hiperparâmetros.

### Interpretação dos Resultados do Tuning

Neste run, o `RandomizedSearchCV` **não produziu uma melhora líquida**: o XGBoost tunado teve F1 ponderado de 0,7677, contra 0,7747 do XGBoost padrão (Δ = −0,0070), e o recall de DESTRUÍDA caiu de 0,38 para 0,26 (Δ = −0,12) — uma perda relevante justamente na classe mais crítica para segurança.

A causa provável: a busca otimiza `f1_weighted` médio via validação cruzada, uma métrica que pondera pelo suporte de cada classe. Como DESTRUÍDA é a classe mais rara (~4% dos dados), configurações que melhoram marginalmente o F1 médio podem, ao mesmo tempo, tornar o modelo mais conservador em relação a essa classe — trocando recall de DESTRUÍDA por precisão/recall nas classes majoritárias. O ganho médio em CV (F1 CV = 0,7800) não se traduziu em ganho no teste, e a classe que mais importa para o objetivo do projeto piorou.

**Decisão:** o **XGBoost padrão (sem tuning)** é mantido como modelo final da Opção B — é ele quem sustenta o F1 = 0,7747 reportado na Seção 7.1. A busca de hiperparâmetros foi executada e documentada de forma metodologicamente correta (sem contaminação do conjunto de teste), mas o resultado concreto deste projeto é que a otimização **não superou o modelo com parâmetros padrão** para o objetivo definido — um achado tão válido quanto uma melhora, e coerente com a limitação estrutural discutida na Seção 7.2: com seis variáveis categóricas e poucos exemplos de DESTRUÍDA, o espaço de ganho via ajuste de algoritmo é estreito.


---

## Seção 7 — Conclusão


### 7.1 — Síntese dos Resultados

| Abordagem | F1 Ponderado |
|---|---|
| Baseline (classe majoritária) | 0,3472 |
| Opção A (modelo por tipo) | 0,44 – 0,81 (varia por tipo) |
| Opção B — XGBoost padrão | **0,7747** |
| Opção B — XGBoost tunado | 0,7677 (não superou o padrão — Seção 6.4) |

A **Opção B com o XGBoost padrão** é a solução recomendada: supera o baseline por larga margem, nunca falha catastroficamente em nenhum segmento, beneficia tipos raros e requer manutenção de um único modelo. A otimização de hiperparâmetros (Seção 6.4) foi executada de forma metodologicamente correta, mas não superou o modelo padrão neste caso — motivo pelo qual o XGBoost sem tuning é o modelo final. O benchmark revelou ainda que acima de ~800 amostras por tipo a Opção A ganha por margem estreita; abaixo de ~400, a Opção B supera por 0,07–0,19 de F1.


### 7.2 — A Limitação Central: Classe DESTRUÍDA

O modelo performa bem em NENHUM (recall ≈ 0,87), SUBSTANCIAL (≈ 0,80) e LEVE (≈ 0,64). O recall de DESTRUÍDA (≈ 0,38) é a principal limitação — e não é corrigível por ajuste de algoritmo ou hiperparâmetros.

DESTRUÍDA representa ~4% do dataset porque décadas de evolução tecnológica e regulatória foram eficazes em reduzir eventos dessa magnitude — modelo de defesas em profundidade de Reason [7]. Ampliar o recorte temporal adicionaria exemplos, mas degradaria a validade semântica: um acidente de 1990 reflete realidade regulatória e tecnológica incompatível com o risco contemporâneo. O recall de 0,38 é o reflexo quantitativo de que a aviação brasileira não produz destruições totais em volume suficiente para aprendizagem supervisionada confiável com seis variáveis categóricas.


### 7.3 — Enriquecimento de Dados: Caminhos Descartados

**`fator_contribuinte`:** taxa de cobertura salta de 7,9% em LEVE para 57,0% em SUBSTANCIAL, mapeando o protocolo institucional do CENIPA. Usar essa variável como feature incorpora o target indiretamente — leakage por variável [4].

**SMOTE:** a interpolação entre valores contínuos produz exemplos sintéticos semanticamente plausíveis em features numéricas [16]. Aplicado a colunas binárias OHE, gera valores fracionários sem interpretação válida. RandomOverSampler duplica as ~548 instâncias de DESTRUÍDA sem introduzir informação nova.


### 7.4 — Ranking de Risco e Trabalhos Futuros

| Tipo | Perfil Observado |
|---|---|
| EXPERIMENTAL | Alta proporção SUBSTANCIAL/DESTRUÍDA; F1 = 0,63 no modelo dedicado |
| AGRÍCOLA | Quatro classes presentes; F1 = 0,77 — consistente com exposição ao voo de baixa altitude |
| POLICIAL | Alto volume amostral, alta variabilidade; F1 = 0,57 |
| REGULAR | Ausência estatística de DESTRUÍDA nos dados recentes |
| PRIVADA / INSTRUÇÃO / TÁXI AÉREO | F1 0,71–0,73; distribuição previsível |


### Trabalhos Futuros

**Próximos passos dentro do escopo estruturado:** (1) calibração de limiar de decisão para DESTRUÍDA (0,25–0,35), sem retreinamento; (2) pesos de classe (`sample_weight`) como alternativa ao oversampling — matematicamente equivalente, sem inflar o dataset.

**Com dados não estruturados:** relatórios narrativos do CENIPA contêm granularidade indisponível nas variáveis estruturadas. O caminho recomendado: TF-IDF sobre seções de fatores causais e embeddings com **BERTimbau** [17]. Esses modelos devem ser enquadrados como ferramentas de *análise retrospectiva*, não de predição em tempo real — os relatórios são produzidos após a determinação da severidade.

### Detecção de Anomalias como Reformulação Alternativa para DESTRUÍDA

A raiz da limitação discutida na Seção 7.2 é estrutural: ~4% de representação é pouco para que um classificador multiclasse aprenda uma fronteira de decisão confiável para DESTRUÍDA via aprendizado supervisionado convencional — a classe simplesmente não fornece exemplos suficientes para o modelo generalizar seu padrão. Uma direção não explorada neste projeto é abandonar a classificação multiclasse para esse subproblema e reformulá-lo como **detecção de anomalias**: em vez de perguntar "a qual das quatro classes esta ocorrência pertence?", perguntar "esta ocorrência se comporta como o regime normal (NENHUM/LEVE/SUBSTANCIAL) ou como um evento atípico?".

Nessa formulação, o modelo aprende a distribuição do regime normal a partir da massa de dados majoritária, e sinaliza como anômalas as ocorrências que se desviam significativamente desse padrão — sem depender de fartura de exemplos da classe rara para definir sua fronteira, que é exatamente a limitação atual. Técnicas candidatas:

- **Isolation Forest** [18]: isola observações via partições aleatórias recursivas, explorando o fato de que anomalias requerem, em média, menos partições para isolamento — adequado ao regime de poucos exemplos positivos, pois não exige exemplos rotulados de DESTRUÍDA para treinar.
- **One-Class SVM**: aprende a fronteira do regime majoritário e classifica como anômalo tudo que cai fora dela.
- **Autoencoders**: aprendem a reconstruir o padrão normal a partir de features estruturadas; um erro de reconstrução elevado sinaliza um evento potencialmente destrutivo.

A troca teria custo: perde-se a granularidade das quatro classes (LEVE e SUBSTANCIAL deixam de ser distinguidas dentro do "não anômalo"), e a saída do modelo passa de "probabilidade por classe" para "score de anomalia", exigindo um novo processo de calibração de limiar operacional. Mas o ganho potencial é direto — modelos de anomalia são desenhados justamente para o regime que hoje é o gargalo do projeto (poucos positivos, muitos negativos) — e essa reformulação seria o teste natural seguinte para isolar se o recall baixo em DESTRUÍDA reflete uma limitação do paradigma de classificação escolhido ou da própria informação disponível nas seis variáveis estruturadas.

### 7.5 — Considerações Finais

Este projeto demonstrou F1 ponderado de **0,7747** com dados estruturados públicos do CENIPA, sem informações produzidas pela investigação posterior. O modelo supera claramente o baseline de classe majoritária e falha de forma previsível e compreendida em DESTRUÍDA — falha investigada, caracterizada e documentada como inerente à estrutura do problema. Em domínios de segurança crítica, documentar o que o modelo *não* consegue fazer é tão relevante quanto reportar o que ele faz bem.


---

## Referências

[1] ICAO. *Annex 13 to the Convention on International Civil Aviation*. 12th ed., 2020.

[2] Li, W.-C.; Harris, D.; Yu, C.-S. "Routes to failure: analysis of 41 civil aviation accidents." *Accident Analysis & Prevention*, v. 40, n. 1, p. 426–434, 2008. https://doi.org/10.1016/j.aap.2007.07.011

[3] CENIPA/FAB. *Anuário Estatístico de Ocorrências Aeronáuticas*. https://www2.fab.mil.br/cenipa/

[4] Kaufman, S. et al. "Leakage in data mining: formulation, detection, and avoidance." *ACM TKDD*, v. 6, n. 4, 2012. https://doi.org/10.1145/2382577.2382579

[5] Tukey, J. W. *Exploratory Data Analysis*. Addison-Wesley, 1977.

[6] Japkowicz, N.; Stephen, S. "The class imbalance problem: a systematic study." *Intelligent Data Analysis*, v. 6, n. 5, 2002. https://doi.org/10.3233/IDA-2002-6504

[7] Reason, J. *Managing the Risks of Organizational Accidents*. Ashgate, 1997.

[8] Sokolova, M.; Lapalme, G. "A systematic analysis of performance measures for classification tasks." *IPM*, v. 45, n. 4, 2009. https://doi.org/10.1016/j.ipm.2009.03.002

[9] Potdar, K. et al. "A comparative study of categorical variable encoding techniques." *IJCA*, v. 175, n. 4, 2017. https://doi.org/10.5120/ijca2017915495

[10] Hastie, T.; Tibshirani, R.; Friedman, J. *The Elements of Statistical Learning*. 2nd ed. Springer, 2009.

[11] Breiman, L. "Random forests." *Machine Learning*, v. 45, n. 1, 2001. https://doi.org/10.1023/A:1010933404324

[12] Chen, T.; Guestrin, C. "XGBoost: a scalable tree boosting system." *KDD*, 2016. https://doi.org/10.1145/2939672.2939785

[13] Fawcett, T. "An introduction to ROC analysis." *Pattern Recognition Letters*, v. 27, n. 8, 2006. https://doi.org/10.1016/j.patrec.2005.10.010

[14] Strobl, C. et al. "Bias in random forest variable importance measures." *BMC Bioinformatics*, v. 8, 2007. https://doi.org/10.1186/1471-2105-8-25

[15] Bergstra, J.; Bengio, Y. "Random search for hyper-parameter optimization." *JMLR*, v. 13, 2012.

[16] Chawla, N. V. et al. "SMOTE: Synthetic Minority Over-sampling Technique." *JAIR*, v. 16, 2002. https://doi.org/10.1613/jair.953

[17] Souza, F. et al. "BERTimbau: pretrained BERT models for Brazilian Portuguese." *BRACIS*, 2020. https://doi.org/10.1007/978-3-030-61377-8_28

[18] Liu, F. T.; Ting, K. M.; Zhou, Z.-H. "Isolation Forest." *IEEE ICDM*, 2008. https://doi.org/10.1109/ICDM.2008.17